### PR TITLE
Shiftblock creation

### PR DESCRIPTION
**Pull request by ssuerbier**

### CODE REVIEW

File: tests/test_scheduling_mip_model.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import pendulum
+
+from staffing.api.rooster_api import OpenShift
+from staffing.shiftblocks.mip_scheduler import MIPScheduler
+
+
+def test():
+    shifts = [
+        OpenShift(1, pendulum.datetime(2005, 7, 14, 0, 0), pendulum.datetime(2005, 7, 14, 1, 0), 1, 0),
```
Comment (ThiagoVieira) [inline tests/test_scheduling_mip_model.py]: OpenShift data class is datetime, while pasing pendulum datetime type object here

File: tests/test_scheduling_mip_model.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        OpenShift(2, pendulum.datetime(2005, 7, 14, 0, 30), pendulum.datetime(2005, 7, 14, 1, 0), 1, 0)
```
Comment (ThiagoVieira) [inline tests/test_scheduling_mip_model.py]: In Openshift data class, might wnt to make slots an optional int and pass None to understand that its an empty data class until filled with a slot

File: staffing/shiftblocks/mip_scheduler.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    def solve(self) -> typing.List[OpenShift]:
+        self._build_model()
+        self._solve_model()
+        return self._transform_solution()
+
+    def _build_model(self):
+        shifts = self.shifts
+
+        # these variables are in class scope for solution retrieval
+        solver = self.solver = pywraplp.Solver("scheduling", pywraplp.Solver.CBC_MIXED_INTEGER_PROGRAMMING)
```
Comment (ThiagoVieira) [inline staffing/shiftblocks/mip_scheduler.py]: Any reason you don't move this to the __init__ part of the class? or in the beginning of solve?

File: staffing/shiftblocks/mip_scheduler.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+    def _transform_solution(self):
+        result = []
+
+        for s in self.shifts:
+            solution_value = self.shift_workers[s.id].SolutionValue()
+
+            if solution_value > 0:
+                s.slots = solution_value
+                result.append(s)
```
Comment (ThiagoVieira) [inline staffing/shiftblocks/mip_scheduler.py]: The optional Int on slots would slightly help here.

File: tests/test_scheduling_mip_model.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import pendulum
+
+from staffing.api.rooster_api import OpenShift
+from staffing.shiftblocks.mip_scheduler import MIPScheduler
+
+
+def test():
```
Comment (ThiagoVieira) [inline tests/test_scheduling_mip_model.py]: Not sure if there are any edge cases you want to test on the constraints, such as testing understaffed/overstaffed penalties) Clearly not to test all the edge cases, but idk if there are frequently hit ones.

File: staffing/shiftblocks/generate_shifts.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+    max_start = ends_at - staffing_rule.min_shift_duration
+
+    for start in pendulum.period(starts_at, max_start).range("minutes", 15):
+        min_end = start + staffing_rule.min_shift_duration
+        max_end = min(start + staffing_rule.max_shift_duration, ends_at)
+
+        for end in pendulum.period(min_end, max_end).range("minutes", 15):
+            if all(check(start, end) for check in constraints):
+                s = OpenShift(count, start, end, starting_point.id, None)
```
Comment (ThiagoVieira) [inline staffing/shiftblocks/generate_shifts.py]: None is used here :)

File: staffing/shiftblocks/generate_shifts.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+
+def _start_between(min, max) -> Callable[[object], bool]:
+    min = min or pendulum.Time(0, 0)
+    max = max or pendulum.Time(23, 45)
+
+    return lambda starts_at, ends_at: min <= starts_at.time() <= max
+
+
+def _covered_by(opening_times: List[OpeningTime]):
```
Comment (ThiagoVieira) [inline staffing/shiftblocks/generate_shifts.py]: Missing a function call? something  Callable[[Tuple[object, object]], bool]: ?

File: staffing/shiftblocks/generate_shifts.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+from typing import List, Callable
+
+import pendulum
+
+from staffing.api.forecast_api import OpeningTime
+from staffing.api.rooster_api import OpenShift, StartingPoint
+
+
+def _as_hours(t: pendulum.Time) -> float:
```
Comment (ThiagoVieira) [inline staffing/shiftblocks/generate_shifts.py]: Is this used? I see the _, but its not being used in this file.

File: staffing/shiftblocks/generate_shifts.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+def _duration_between(min, max) -> Callable[[object], bool]:
+    min = min or pendulum.duration()
+    max = max or pendulum.duration(hours=12)
+
+    return lambda starts_at, ends_at: min <= (ends_at - starts_at) <= max
+
+
+def _start_between(min, max) -> Callable[[object], bool]:
+    min = min or pendulum.Time(0, 0)
```
Comment (ThiagoVieira) [inline staffing/shiftblocks/generate_shifts.py]: ignore this coment

File: staffing/shiftblocks/generate_shifts.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+def _as_hours(t: pendulum.Time) -> float:
+    return t.hour + t.minute / 60.0
+
+
+def _duration_between(min, max) -> Callable[[object], bool]:
+    min = min or pendulum.duration()
+    max = max or pendulum.duration(hours=12)
+
+    return lambda starts_at, ends_at: min <= (ends_at - starts_at) <= max
```
Comment (ThiagoVieira) [inline staffing/shiftblocks/generate_shifts.py]: This looks weird, a 0 duration min max could pass on default?  Wouldn't be hit on the for loop, but jw

File: tests/test_generate_shifts.py [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+
+def test_max_start():
+    start = pendulum.datetime(2017, 1, 1, 16, 0)
+    end = pendulum.datetime(2017, 1, 2)
+    starting_point = _create_starting_point(pendulum.duration(hours=2), pendulum.duration(hours=2),
+                                            pendulum.time(10, 0), pendulum.time(20, 0))
+    shifts = generate_shifts.generate_shifts(start, end, starting_point, opening_times)
+
+    assert all(s.starts_at.time() <= pendulum.time(20, 0) for s in shifts)
```
Comment (ThiagoVieira) [inline tests/test_generate_shifts.py]: maybe also check number of shifts that go through properly? (17?)

File: tests/test_generate_shifts.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+
+
+def test_duration():
+    start = pendulum.datetime(2017, 1, 1, 16, 0)
+    end = pendulum.datetime(2017, 1, 1, 20, 0)
+    starting_point = _create_starting_point(pendulum.duration(hours=2), pendulum.duration(hours=2),
+                                            pendulum.time(10, 0), pendulum.time(20, 0))
+    shifts = generate_shifts.generate_shifts(start, end, starting_point, opening_times)
+
+    assert all((s.ends_at - s.starts_at).as_interval() == pendulum.duration(hours=2) for s in shifts)
```
Comment (ThiagoVieira) [inline tests/test_generate_shifts.py]: maybe also check number of shifts that go through properly?

File: staffing/api/forecast_api.py [hunk 13]
```diff
@@ -26,8 +26,8 @@ class ForecastService:
     """Client for Forecast API to fetch opening times"""
     client: APIClient
 
-    def __init__(self, country, token):
-        self.client = APIClient(f"https://{country}.usehurrier.com/api/forecast/v1/", token)
+    def __init__(self, url, token):
+        self.client = APIClient(url, token)
```
Comment (ThiagoVieira) [inline staffing/api/forecast_api.py]: The Types in def create_parallel and delete_parallel don't seem to match whats declared.

File: staffing/api/forecast_api.py [hunk 14]
```diff
@@ -26,8 +26,8 @@ class ForecastService:
     """Client for Forecast API to fetch opening times"""
     client: APIClient
 
-    def __init__(self, country, token):
-        self.client = APIClient(f"https://{country}.usehurrier.com/api/forecast/v1/", token)
+    def __init__(self, url, token):
```
Comment (ThiagoVieira) [inline staffing/api/forecast_api.py]: Also, in the def get(self, endpoint: str, entity_id: int, filter={}): you shouldn't pass an empty dictionary (or list) as a default starting argument to a function, as every invocation will point to the same empty dictionary, a new empty dictionary will not get passed each time (I believe).

Comment (ThiagoVieira) [reply staffing/api/forecast_api.py]: And in def list(..)

File: tests/test_scheduling_mip_model.py (hunk 1 above)
Comment (ssuerbier) [reply tests/test_scheduling_mip_model.py]: Good spot, changed it in `OpenShift` to `pendulum.DateTime`)

File: tests/test_scheduling_mip_model.py (hunk 2 above)
Comment (ssuerbier) [reply tests/test_scheduling_mip_model.py]: :+1:

File: staffing/shiftblocks/mip_scheduler.py (hunk 3 above)
Comment (ssuerbier) [reply staffing/shiftblocks/mip_scheduler.py]: Reason yes, maybe weak. The solver is only used internally (int `_transform_solution`) and I wanted to hide it. Maybe the convention of prepending the underscore is a better choice, wdyt?

File: staffing/api/forecast_api.py (hunk 13 above)
Comment (ssuerbier) [reply staffing/api/forecast_api.py]: This is the type checker being confused. Let me try to fix that, but the session for the parallel stuff comes from request_futures which returns `Future` for all requests.

Comment (ssuerbier) [reply staffing/api/forecast_api.py]: Ah ok I know why, because `FuturesSession` extends the normal requests `Session` and the typechecker doesnt understand this :|

File: staffing/api/forecast_api.py (hunk 14 above)
Comment (ssuerbier) [reply staffing/api/forecast_api.py]: Good point.
